### PR TITLE
Draft/lw/callbacks

### PR DESCRIPTION
**Pull request by stramit**

### CODE REVIEW

File: com.unity.render-pipelines.lightweight/LWRP/Passes/LwForwardPass.cs [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+
+
+        Material m_ErrorMaterial;
+
+        MixedLightingSetup m_MixedLightingSetup;
+        List<ShaderPassName> m_LegacyShaderPassNames;
+        protected RendererConfiguration rendererConfiguration;
+        protected bool dynamicBatching;
+
+        protected LwForwardPass(LightweightForwardRenderer renderer) : base(renderer)
```
Comment (phi-lira) [inline com.unity.render-pipelines.lightweight/LWRP/Passes/LwForwardPass.cs]: I have a concern here. RenderOpaques and RenderTransparents inherit from ForwardPass to setup lights constant and RT. This way we set light constants multiple times. I propose to convert this to SetupLightConstants pass and delegate the RenderTarget setup to opaques and transparents passes.

Comment (phi-lira) [reply com.unity.render-pipelines.lightweight/LWRP/Passes/LwForwardPass.cs]: We can also merge this SetupLightConstants with the SetupForwardRendering pass we already have.

Comment (stramit) [reply com.unity.render-pipelines.lightweight/LWRP/Passes/LwForwardPass.cs]: Yeah I was thinking about this one also. I think what you say makes sense.

File: com.unity.render-pipelines.lightweight/LWRP/DefaultLWRenderer.cs [hunk 2]
```diff
@@ -0,0 +1,184 @@
+using UnityEngine.Rendering;
+
+namespace UnityEngine.Experimental.Rendering.LightweightPipeline
+{
+    public class DefaultLWRenderer : IRendererSetup
```
[the remark this answers is not in the document]
Comment (stramit) [reply com.unity.render-pipelines.lightweight/LWRP/DefaultLWRenderer.cs]: DefaultRendererSetup, it's in the LW namespace

File: com.unity.render-pipelines.lightweight/LWRP/LightweightPipeline.cs [hunk 3]
```diff
@@ -112,7 +118,11 @@ public override void Render(ScriptableRenderContext context, Camera[] cameras)
                         InitializeRenderingData(ref cameraData, visibleLights,
                             m_Renderer.maxSupportedLocalLightsPerPass, m_Renderer.maxSupportedVertexLights,
                             out renderingData);
-                        m_Renderer.Setup(ref context, ref m_CullResults, ref renderingData);
+
+                        var setup = cameraData.camera.GetComponent<IRendererSetup>();
+                        if (setup == null)
+                            setup = new DefaultLWRenderer();
```
[the remark this answers is not in the document]
Comment (stramit) [reply com.unity.render-pipelines.lightweight/LWRP/LightweightPipeline.cs]: done

File: com.unity.render-pipelines.lightweight/LWRP/DefaultLWRenderer.cs [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+            return clearFlag;
+        }
+
+        RendererConfiguration GetRendererConfiguration(int localLightsCount)
+        {
+            RendererConfiguration configuration = RendererConfiguration.PerObjectReflectionProbes | RendererConfiguration.PerObjectLightmaps | RendererConfiguration.PerObjectLightProbe;
+            if (localLightsCount > 0)
+            {
+                if ( m_Renderer.useComputeBufferForPerObjectLightIndices)
```
[the remark this answers is not in the document]
Comment (stramit) [reply com.unity.render-pipelines.lightweight/LWRP/DefaultLWRenderer.cs]: done

File: com.unity.render-pipelines.lightweight/LWRP/DefaultLWRenderer.cs [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        private RenderOpaqueForward m_RenderOpaqueForward;
+        private OpaquePostProcessPass m_OpaquePostProcessPass;
+        private CopyDepthPass m_CopyDepthPass;
+        private CopyColorPass m_CopyColorPass;
+        private RenderTransparentForward m_RenderTransparentForward;
+        private TransparentPostProcessPass m_TransparentPostProcessPass;
+        private FinalBlitPass m_FinalBlitPass;
+        private EndXRRendering m_EndXrRendering;
+
+        private LightweightForwardRenderer m_Renderer;
```
[the remark this answers is not in the document]
Comment (stramit) [reply com.unity.render-pipelines.lightweight/LWRP/DefaultLWRenderer.cs]: done

File: com.unity.render-pipelines.lightweight/LWRP/LightweightPipeline.cs [hunk 6]
```diff
@@ -378,7 +388,7 @@ void SetupPerCameraShaderConstants(CameraData cameraData)
         {
             float cameraWidth = (float)cameraData.camera.pixelWidth * cameraData.renderScale;
             float cameraHeight = (float)cameraData.camera.pixelWidth * cameraData.renderScale;
-            Shader.SetGlobalVector(PerCameraBuffer._ScaledScreenParams, new Vector4(cameraWidth, cameraHeight, 1.0f + 1.0f / cameraWidth, 1.0f + 1.0f / cameraHeight));
+            Shader.SetGlobalVector(LwForwardPass.PerCameraBuffer._ScaledScreenParams, new Vector4(cameraWidth, cameraHeight, 1.0f + 1.0f / cameraWidth, 1.0f + 1.0f / cameraHeight));
```
[the remark this answers is not in the document]
Comment (stramit) [reply com.unity.render-pipelines.lightweight/LWRP/LightweightPipeline.cs]: I'll leave this one for a future tidy up

File: com.unity.render-pipelines.lightweight/LWRP/Passes/ConfigureForwardRTs.cs [hunk 7]
```diff
@@ -0,0 +1,69 @@
+using UnityEngine.Rendering;
+
+namespace UnityEngine.Experimental.Rendering.LightweightPipeline
+{
+    public class ConfigureForwardRTs : ScriptableRenderPass
+    {
+        public ConfigureForwardRTs(LightweightForwardRenderer renderer) : base(renderer)
```
[the remark this answers is not in the document]
Comment (stramit) [reply com.unity.render-pipelines.lightweight/LWRP/Passes/ConfigureForwardRTs.cs]: done

File: com.unity.render-pipelines.lightweight/LWRP/DefaultLWRenderer.cs [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+        private bool m_Initialized = false;
+
+        private void Init(LightweightForwardRenderer renderer)
```
[the remark this answers is not in the document]
Comment (stramit) [reply com.unity.render-pipelines.lightweight/LWRP/DefaultLWRenderer.cs]: Agree, I'll leave that for future work

File: com.unity.render-pipelines.lightweight/LWRP/LightweightForwardRenderer.cs [hunk 9]
```diff
@@ -26,15 +17,63 @@ public enum MaterialHandles
         Count,
     }
 
+    public struct RenderTargetHandle
```
[the remark this answers is not in the document]
Comment (stramit) [reply com.unity.render-pipelines.lightweight/LWRP/LightweightForwardRenderer.cs]: Moved them to their own files already.

File: com.unity.render-pipelines.lightweight/LWRP/Passes/CopyDepthPass.cs [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+{
+    public class CopyDepthPass : ScriptableRenderPass
+    {
+        Material m_DepthCopyMaterial;
+
+        private RenderTargetHandle depthAttachmentHandle { get; set; }
+
+        public CopyDepthPass(LightweightForwardRenderer renderer) : base(renderer)
+        {
+            // Copy Depth Pass
```
[the remark this answers is not in the document]
Comment (stramit) [reply com.unity.render-pipelines.lightweight/LWRP/Passes/CopyDepthPass.cs]: done

File: com.unity.render-pipelines.lightweight/LWRP/Passes/FinalBlitPass.cs [hunk 11]
```diff
@@ -0,0 +1,57 @@
+using UnityEngine.Rendering;
+
+namespace UnityEngine.Experimental.Rendering.LightweightPipeline
+{
+    public class FinalBlitPass : ScriptableRenderPass
+    {
+        //RenderTextureFormat m_ColorFormat;
```
[the remark this answers is not in the document]
Comment (stramit) [reply com.unity.render-pipelines.lightweight/LWRP/Passes/FinalBlitPass.cs]: done

File: com.unity.render-pipelines.lightweight/LWRP/Passes/LwForwardPass.cs (hunk 1 above)
Comment (stramit) [reply com.unity.render-pipelines.lightweight/LWRP/Passes/LwForwardPass.cs]: I think I might leave this change to you as it's not just as simple as pulling it out in a nice way.

File: com.unity.render-pipelines.lightweight/LWRP/DefaultLWRenderer.cs [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+                renderer.EnqueuePass(m_FinalBlitPass);
+            }
+
+            if (renderingData.cameraData.isStereoEnabled)
+            {
+                renderer.EnqueuePass(m_EndXrRendering);
+            }
+        } 
+
+        private static ClearFlag GetCameraClearFlag(Camera camera)
```
[the remark this answers is not in the document]
Comment (stramit) [reply com.unity.render-pipelines.lightweight/LWRP/DefaultLWRenderer.cs]: done

**Pull request merged**